File: programmers/154540.js
const steps = [
  [1, 0],
  [-1, 0],
  [0, 1],
  [0, -1],
];

const collect = (graph, visited, start) => {
  let collected = 0;

  const [x, y] = start;
  if (visited[x][y] || graph[x][y] === 0) return 0;

  const stack = [start];
  visited[start[0]][start[1]] = true;

  while (stack.length > 0) {
    const [curX, curY] = stack.pop();
    collected += graph[curX][curY];

    for (const step of steps) {
      const nextX = curX + step[0];
      const nextY = curY + step[1];

      if (nextX < 0 || nextX > graph.length - 1 || nextY < 0 || nextY > graph[0].length - 1) {
        continue;
      }

      if (visited[nextX][nextY] || graph[nextX][nextY] === 0) {
        continue;
      }

      stack.push([nextX, nextY]);
      visited[nextX][nextY] = true;
    }
  }

  return collected;
};

function solution(maps) {
  const graph = maps.map(row => row.split('').map(v => (v === 'X' ? 0 : Number(v))));
  const answer = [];

  const visited = Array(graph.length)
    .fill(null)
    .map(_ => Array(graph[0].length).fill(false));

  for (let x = 0; x < graph.length; x++) {
    for (let y = 0; y < graph[0].length; y++) {
      answer.push(collect(graph, visited, [x, y]));
    }
  }

  const validAnswer = answer.filter(x => x > 0).sort((a, b) => a - b);

  return validAnswer.length === 0 ? [-1] : validAnswer;
}

console.log(solution(['X591X', 'X1X5X', 'X231X', '1XXX1']));
console.log(solution(['XXX', 'XXX', 'XXX']));

File: programmers/154540.py
from collections import deque
from typing import List, Tuple


def solution(maps: List[str]) -> List[int]:
    answer = []
    visited = [[False] * len(maps[0]) for _ in range(len(maps))]

    for i in range(len(maps)):
        for j in range(len(maps[0])):
            if maps[i][j] != "X" and visited[i][j] is not True:
                answer.append(bfs(maps, visited, (i, j)))

    if len(answer) <= 0:
        return [-1]

    answer.sort()
    return answer


def bfs(maps: List[str], visited: List[List[bool]], start: Tuple[int, int]) -> int:
    queue = deque([start])
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    food_cnt = 0
    visited[start[0]][start[1]] = True

    while queue:
        cur_x, cur_y = queue.popleft()
        food_cnt += int(maps[cur_x][cur_y])

        for i in range(4):
            next_x = cur_x + dx[i]
            next_y = cur_y + dy[i]

            if not (0 <= next_x < len(maps)) or not (0 <= next_y < len(maps[0])):
                continue
            if maps[next_x][next_y] == "X":
                continue
            if visited[next_x][next_y]:
                continue

            visited[next_x][next_y] = True
            queue.append((next_x, next_y))

    return food_cnt


print(solution(["X591X", "X1X5X", "X231X", "1XXX1"]))
